File: my-birds/bird-watching-frontend/src/tests/geolocation-integration.test.ts
/**
 * Integration test for geolocation feature
 * This test validates the complete frontend flow of the geolocation feature:
 * 1. Coordinate validation utilities
 * 2. Coordinate formatting
 * 3. Map data filtering
 * 4. Geolocation service
 * 5. Complete observation flow with coordinates
 */

import { describe, it, expect } from 'vitest';
import {
  isValidLatitude,
  isValidLongitude,
  formatCoordinate,
  formatCoordinateWithDirection,
  formatForGPS,
} from '../utils/coordinateUtils';

// Helper function to validate coordinate pairs
function validateCoordinatePair(lat: number, lng: number): boolean {
  return isValidLatitude(lat) && isValidLongitude(lng);
}

describe('Geolocation Feature Integration Tests', () => {
  describe('Step 1: Coordinate Validation', () => {
    it('should validate latitude bounds correctly', () => {
      console.log('✓ Testing latitude validation...');
      
      // Valid latitudes
      expect(isValidLatitude(0)).toBe(true);
      expect(isValidLatitude(90)).toBe(true);
      expect(isValidLatitude(-90)).toBe(true);
      expect(isValidLatitude(45.5)).toBe(true);
      expect(isValidLatitude(-45.5)).toBe(true);

      // Invalid latitudes
      expect(isValidLatitude(90.1)).toBe(false);
      expect(isValidLatitude(-90.1)).toBe(false);
      expect(isValidLatitude(95)).toBe(false);
      expect(isValidLatitude(-95)).toBe(false);
      expect(isValidLatitude(NaN)).toBe(false);
      expect(isValidLatitude(Infinity)).toBe(false);
      
      console.log('  ✓ Latitude validation works correctly');
    });

    it('should validate longitude bounds correctly', () => {
      console.log('✓ Testing longitude validation...');
      
      // Valid longitudes
      expect(isValidLongitude(0)).toBe(true);
      expect(isValidLongitude(180)).toBe(true);
      expect(isValidLongitude(-180)).toBe(true);
      expect(isValidLongitude(120.5)).toBe(true);
      expect(isValidLongitude(-120.5)).toBe(true);

      // Invalid longitudes
      expect(isValidLongitude(180.1)).toBe(false);
      expect(isValidLongitude(-180.1)).toBe(false);
      expect(isValidLongitude(200)).toBe(false);
      expect(isValidLongitude(-200)).toBe(false);
      expect(isValidLongitude(NaN)).toBe(false);
      expect(isValidLongitude(Infinity)).toBe(false);
      
      console.log('  ✓ Longitude validation works correctly');
    });

    it('should validate coordinate pairs correctly', () => {
      console.log('✓ Testing coordinate pair validation...');
      
      // Valid pairs
      expect(validateCoordinatePair(40.7128, -74.0060)).toBe(true); // New York
      expect(validateCoordinatePair(51.5074, -0.1278)).toBe(true);  // London
      expect(validateCoordinatePair(-33.8688, 151.2093)).toBe(true); // Sydney
      expect(validateCoordinatePair(0, 0)).toBe(true);               // Null Island
      expect(validateCoordinatePair(90, 180)).toBe(true);            // Boundaries
      expect(validateCoordinatePair(-90, -180)).toBe(true);          // Boundaries

      // Invalid pairs
      expect(validateCoordinatePair(95, -74.0060)).toBe(false);     // Invalid latitude
      expect(validateCoordinatePair(40.7128, 200)).toBe(false);     // Invalid longitude
      expect(validateCoordinatePair(-100, 0)).toBe(false);          // Invalid latitude
      expect(validateCoordinatePair(0, -200)).toBe(false);          // Invalid longitude
      expect(validateCoordinatePair(NaN, 0)).toBe(false);           // NaN latitude
      expect(validateCoordinatePair(0, NaN)).toBe(false);           // NaN longitude
      
      console.log('  ✓ Coordinate pair validation works correctly');
    });
  });

  describe('Step 2: Coordinate Formatting', () => {
    it('should format coordinates with correct precision', () => {
      console.log('✓ Testing coordinate formatting...');
      
      // Test latitude formatting
      expect(formatCoordinate(40.7128, 'lat')).toBe('40.712800');
      expect(formatCoordinate(-33.8688, 'lat')).toBe('-33.868800');
      expect(formatCoordinate(0, 'lat')).toBe('0.000000');
      expect(formatCoordinate(90, 'lat')).toBe('90.000000');
      expect(formatCoordinate(-90, 'lat')).toBe('-90.000000');

      // Test longitude formatting
      expect(formatCoordinate(-74.0060, 'lng')).toBe('-74.006000');
      expect(formatCoordinate(151.2093, 'lng')).toBe('151.209300');
      expect(formatCoordinate(0, 'lng')).toBe('0.000000');
      expect(formatCoordinate(180, 'lng')).toBe('180.000000');
      expect(formatCoordinate(-180, 'lng')).toBe('-180.000000');
      
      console.log('  ✓ Coordinate formatting works correctly');
    });

    it('should format coordinates with directional indicators', () => {
      console.log('✓ Testing directional formatting...');
      
      // Test all quadrants
      expect(formatCoordinateWithDirection(40.7128, -74.0060)).toBe('40.712800°N, 74.006000°W');
      expect(formatCoordinateWithDirection(51.5074, -0.1278)).toBe('51.507400°N, 0.127800°W');
      expect(formatCoordinateWithDirection(-33.8688, 151.2093)).toBe('33.868800°S, 151.209300°E');
      expect(formatCoordinateWithDirection(-34.6037, -58.3816)).toBe('34.603700°S, 58.381600°W');
      
      // Test zero coordinates
      expect(formatCoordinateWithDirection(0, 0)).toBe('0.000000°N, 0.000000°E');
      
      // Test boundaries
      expect(formatCoordinateWithDirection(90, 180)).toBe('90.000000°N, 180.000000°E');
      expect(formatCoordinateWithDirection(-90, -180)).toBe('90.000000°S, 180.000000°W');
      
      console.log('  ✓ Directional formatting works correctly');
    });

    it('should format coordinates for GPS devices', () => {
      console.log('✓ Testing GPS formatting...');
      
      // Test GPS format (decimal degrees)
      expect(formatForGPS(40.7128, -74.0060)).toBe('40.712800, -74.006000');
      expect(formatForGPS(51.5074, -0.1278)).toBe('51.507400, -0.127800');
      expect(formatForGPS(-33.8688, 151.2093)).toBe('-33.868800, 151.209300');
      expect(formatForGPS(0, 0)).toBe('0.000000, 0.000000');
      
      console.log('  ✓ GPS formatting works correctly');
    });
  });

  describe('Step 3: Map Data Filtering', () => {
    it('should filter observations with coordinates', () => {
      console.log('✓ Testing observation filtering...');
      
      const observations = [
        { id: '1', species_name: 'Cardinal', latitude: 40.7128, longitude: -74.0060 },
        { id: '2', species_name: 'Blue Jay', latitude: null, longitude: null },
        { id: '3', species_name: 'Robin', latitude: 51.5074, longitude: -0.1278 },
        { id: '4', species_name: 'Sparrow', latitude: undefined, longitude: undefined },
        { id: '5', species_name: 'Hawk', latitude: -33.8688, longitude: 151.2093 },
      ];

      const withCoordinates = observations.filter(
        obs => obs.latitude != null && obs.longitude != null
      );

      expect(withCoordinates).toHaveLength(3);
      expect(withCoordinates.map(o => o.id)).toEqual(['1', '3', '5']);
      
      console.log('  ✓ Observation filtering works correctly');
      console.log(`    Found ${withCoordinates.length} observations with coordinates`);
    });

    it('should filter shared observations with coordinates', () => {
      console.log('✓ Testing shared observation filtering...');
      
      const observations = [
        { id: '1', species_name: 'Cardinal', is_shared: true, latitude: 40.7128, longitude: -74.0060 },
        { id: '2', species_name: 'Blue Jay', is_shared: false, latitude: 51.5074, longitude: -0.1278 },
        { id: '3', species_name: 'Robin', is_shared: true, latitude: null, longitude: null },
        { id: '4', species_name: 'Sparrow', is_shared: true, latitude: -33.8688, longitude: 151.2093 },
      ];

      const sharedWithCoordinates = observations.filter(
        obs => obs.is_shared && obs.latitude != null && obs.longitude != null
      );

      expect(sharedWithCoordinates).toHaveLength(2);
      expect(sharedWithCoordinates.map(o => o.id)).toEqual(['1', '4']);
      
      console.log('  ✓ Shared observation filtering works correctly');
      console.log(`    Found ${sharedWithCoordinates.length} shared observations with coordinates`);
    });

    it('should filter trip observations with coordinates', () => {
      console.log('✓ Testing trip observation filtering...');
      
      const tripId = 'trip-123';
      const observations = [
        { id: '1', species_name: 'Cardinal', trip_id: tripId, latitude: 40.7128, longitude: -74.0060 },
        { id: '2', species_name: 'Blue Jay', trip_id: tripId, latitude: null, longitude: null },
        { id: '3', species_name: 'Robin', trip_id: 'other-trip', latitude: 51.5074, longitude: -0.1278 },
        { id: '4', species_name: 'Sparrow', trip_id: tripId, latitude: -33.8688, longitude: 151.2093 },
      ];

      const tripObservationsWithCoordinates = observations.filter(
        obs => obs.trip_id === tripId && obs.latitude != null && obs.longitude != null
      );

      expect(tripObservationsWithCoordinates).toHaveLength(2);
      expect(tripObservationsWithCoordinates.map(o => o.id)).toEqual(['1', '4']);
      
      console.log('  ✓ Trip observation filtering works correctly');
      console.log(`    Found ${tripObservationsWithCoordinates.length} trip observations with coordinates`);
    });
  });

  describe('Step 4: Complete Observation Flow', () => {
    it('should handle observation creation with coordinates', () => {
      console.log('✓ Testing observation creation flow...');
      
      // Simulate form data
      const formData = {
        species_name: 'Northern Cardinal',
        observation_date: new Date().toISOString(),
        location: 'Central Park, New York',
        latitude: 40.785091,
        longitude: -73.968285,
        notes: 'Beautiful red bird spotted near the lake',
        is_shared: true,
      };

      // Validate coordinates
      expect(isValidLatitude(formData.latitude)).toBe(true);
      expect(isValidLongitude(formData.longitude)).toBe(true);
      expect(validateCoordinatePair(formData.latitude, formData.longitude)).toBe(true);

      // Format for display
      const displayCoords = formatCoordinateWithDirection(formData.latitude, formData.longitude);
      expect(displayCoords).toBe('40.785091°N, 73.968285°W');

      // Format for GPS
      const gpsCoords = formatForGPS(formData.latitude, formData.longitude);
      expect(gpsCoords).toBe('40.785091, -73.968285');

      console.log('  ✓ Observation creation flow works correctly');
      console.log(`    Species: ${formData.species_name}`);
      console.log(`    Location: ${formData.location}`);
      console.log(`    Coordinates: ${displayCoords}`);
    });

    it('should handle observation creation without coordinates', () => {
      console.log('✓ Testing observation creation without coordinates...');
      
      // Simulate form data without coordinates
      const formData = {
        species_name: 'Blue Jay',
        observation_date: new Date().toISOString(),
        location: 'Somewhere in the park',
        latitude: undefined,
        longitude: undefined,
        notes: null,
        is_shared: false,
      };

      // Verify coordinates are optional
      expect(formData.latitude).toBeUndefined();
      expect(formData.longitude).toBeUndefined();

      console.log('  ✓ Observation without coordinates handled correctly');
    });

    it('should reject invalid coordinates in observation creation', () => {
      console.log('✓ Testing coordinate validation in observation creation...');
      
      const testCases = [
        {
          name: 'Invalid latitude',
          data: { latitude: 95, longitude: -74.0060 },
          expectedError: 'latitude',
        },
        {
          name: 'Invalid longitude',
          data: { latitude: 40.7128, longitude: 200 },
          expectedError: 'longitude',
        },
        {
          name: 'Incomplete coordinates (only latitude)',
          data: { latitude: 40.7128, longitude: undefined },
          expectedError: 'both',
        },
        {
          name: 'Incomplete coordinates (only longitude)',
          data: { latitude: undefined, longitude: -74.0060 },
          expectedError: 'both',
        },
      ];

      testCases.forEach(({ name, data, expectedError }) => {
        const { latitude, longitude } = data;
        
        let isValid = true;
        let errorType = '';

        // Check if coordinates are provided
        if (latitude !== undefined || longitude !== undefined) {
          // Check if both are provided
          if (latitude === undefined || longitude === undefined) {
            isValid = false;
            errorType = 'both';
          } else {
            // Validate bounds
            if (!isValidLatitude(latitude)) {
              isValid = false;
              errorType = 'latitude';
            }
            if (!isValidLongitude(longitude)) {
              isValid = false;
              errorType = 'longitude';
            }
          }
        }

        expect(isValid).toBe(false);
        expect(errorType).toBe(expectedError);
        console.log(`  ✓ ${name} rejected correctly`);
      });
    });

    it('should handle coordinate updates', () => {
      console.log('✓ Testing coordinate updates...');
      
      // Initial observation
      const observation = {
        id: '123',
        species_name: 'Northern Cardinal',
        location: 'Central Park, New York',
        latitude: 40.785091,
        longitude: -73.968285,
      };

      console.log(`  Initial location: ${observation.location}`);
      console.log(`  Initial coordinates: ${formatCoordinateWithDirection(observation.latitude, observation.longitude)}`);

      // Update coordinates
      const updates = {
        location: 'Times Square, New York',
        latitude: 40.758896,
        longitude: -73.985130,
      };

      // Validate new coordinates
      expect(isValidLatitude(updates.latitude)).toBe(true);
      expect(isValidLongitude(updates.longitude)).toBe(true);
      expect(validateCoordinatePair(updates.latitude, updates.longitude)).toBe(true);

      // Apply updates
      const updatedObservation = { ...observation, ...updates };

      expect(updatedObservation.latitude).toBe(40.758896);
      expect(updatedObservation.longitude).toBe(-73.985130);
      expect(updatedObservation.location).toBe('Times Square, New York');

      console.log(`  Updated location: ${updatedObservation.location}`);
      console.log(`  Updated coordinates: ${formatCoordinateWithDirection(updatedObservation.latitude, updatedObservation.longitude)}`);
      console.log('  ✓ Coordinate updates work correctly');
    });
  });

  describe('Step 5: Proximity Search Simulation', () => {
    it('should calculate distances for proximity search', () => {
      console.log('✓ Testing proximity search distance calculations...');
      
      // Simple distance calculation (Haversine formula)
      const haversineDistance = (lat1: number, lon1: number, lat2: number, lon2: number): number => {
        const R = 6371; // Earth's radius in km
        const dLat = (lat2 - lat1) * Math.PI / 180;
        const dLon = (lon2 - lon1) * Math.PI / 180;
        const a = 
          Math.sin(dLat / 2) * Math.sin(dLat / 2) +
          Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
          Math.sin(dLon / 2) * Math.sin(dLon / 2);
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
      };

      // Times Square coordinates
      const centerLat = 40.758896;
      const centerLng = -73.985130;

      // Test observations
      const observations = [
        { id: '1', name: 'Times Square', lat: 40.758896, lng: -73.985130 },
        { id: '2', name: 'Brooklyn Bridge', lat: 40.706086, lng: -73.996864 },
        { id: '3', name: 'Statue of Liberty', lat: 40.689247, lng: -74.044502 },
      ];

      // Calculate distances
      const withDistances = observations.map(obs => ({
        ...obs,
        distance: haversineDistance(centerLat, centerLng, obs.lat, obs.lng),
      }));

      // Verify distances
      expect(withDistances[0].distance).toBeLessThan(0.1); // Same location
      expect(withDistances[1].distance).toBeGreaterThan(5);
      expect(withDistances[1].distance).toBeLessThan(7);
      expect(withDistances[2].distance).toBeGreaterThan(9);
      expect(withDistances[2].distance).toBeLessThan(11);

      console.log('  Distances from Times Square:');
      withDistances.forEach(obs => {
        console.log(`    ${obs.name}: ${obs.distance.toFixed(2)} km`);
      });

      // Test proximity filtering (5 km radius)
      const radius = 5;
      const nearby = withDistances.filter(obs => obs.distance <= radius);
      
      expect(nearby).toHaveLength(1);
      expect(nearby[0].id).toBe('1');
      
      console.log(`  ✓ Found ${nearby.length} observations within ${radius} km`);

      // Test proximity filtering (15 km radius)
      const largerRadius = 15;
      const nearbyLarger = withDistances.filter(obs => obs.distance <= largerRadius);
      
      expect(nearbyLarger).toHaveLength(3);
      
      console.log(`  ✓ Found ${nearbyLarger.length} observations within ${largerRadius} km`);
    });
  });

  describe('Step 6: Summary', () => {
    it('should summarize all tested features', () => {
      console.log('\n=== ✅ All Frontend Integration Tests Passed! ===\n');
      console.log('Tested features:');
      console.log('  ✓ Coordinate validation (latitude bounds)');
      console.log('  ✓ Coordinate validation (longitude bounds)');
      console.log('  ✓ Coordinate validation (pair requirement)');
      console.log('  ✓ Coordinate formatting (precision)');
      console.log('  ✓ Coordinate formatting (directional indicators)');
      console.log('  ✓ Coordinate formatting (GPS format)');
      console.log('  ✓ Map data filtering (observations with coordinates)');
      console.log('  ✓ Map data filtering (shared observations)');
      console.log('  ✓ Map data filtering (trip observations)');
      console.log('  ✓ Observation creation with coordinates');
      console.log('  ✓ Observation creation without coordinates');
      console.log('  ✓ Coordinate validation in forms');
      console.log('  ✓ Coordinate updates');
      console.log('  ✓ Proximity search distance calculations');
      console.log('  ✓ Proximity search filtering\n');
      
      expect(true).toBe(true);
    });
  });
});

File: my-birds/bird-watching-backend/tests/geolocation_integration_test.rs
// Integration test for geolocation feature
// This test validates the complete flow of the geolocation feature:
// 1. Create observation with coordinates
// 2. Validate coordinate storage and retrieval
// 3. Test proximity search
// 4. Test coordinate validation
// 5. Test coordinate updates

use bird_watching_backend::models::observation::{CreateObservationRequest, UpdateObservationRequest};
use bird_watching_backend::models::user::RegisterRequest;
use bird_watching_backend::services::auth_service::AuthService;
use bird_watching_backend::services::observation_service::ObservationService;
use bird_watching_backend::services::geo_service::GeoService;
use chrono::{Duration, Utc};
use sqlx::postgres::PgPoolOptions;
use std::env;
use uuid::Uuid;

// Helper function to get test database pool
async fn get_test_pool() -> sqlx::PgPool {
    let database_url = env::var("DATABASE_URL").unwrap_or_else(|_| {
        "postgresql://postgres:password@localhost:5432/bird_watching".to_string()
    });

    PgPoolOptions::new()
        .max_connections(5)
        .connect(&database_url)
        .await
        .expect("Failed to create test pool")
}

// Helper function to clean up test data
async fn cleanup_user(pool: &sqlx::PgPool, username: &str) {
    let _ = sqlx::query("DELETE FROM users WHERE username = $1")
        .bind(username)
        .execute(pool)
        .await;
}

#[tokio::test]
async fn test_complete_geolocation_flow() {
    let pool = get_test_pool().await;
    let auth_service = AuthService::new(pool.clone());
    let observation_service = ObservationService::new(pool.clone());

    // Generate unique test data
    let uuid_suffix = Uuid::new_v4().to_string().replace("-", "")[..8].to_string();
    let username = format!("testuser_{}", uuid_suffix);
    let email = format!("test_{}@example.com", uuid_suffix);
    let password = "TestPassword123!";

    // Clean up any existing test data
    cleanup_user(&pool, &username).await;

    println!("=== Step 1: Register test user ===");
    let user = auth_service
        .register(RegisterRequest {
            username: username.clone(),
            email: email.clone(),
            password: password.to_string(),
        })
        .await
        .expect("User registration should succeed");
    
    println!("✓ User registered: {} ({})", user.username, user.id);

    // Test coordinates (Central Park, New York)
    let central_park_lat = 40.785091;
    let central_park_lng = -73.968285;

    println!("\n=== Step 2: Create observation with coordinates ===");
    let observation_date = Utc::now() - Duration::days(1);
    let observation = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "Northern Cardinal".to_string(),
                observation_date,
                location: "Central Park, New York".to_string(),
                latitude: Some(central_park_lat),
                longitude: Some(central_park_lng),
                notes: Some("Beautiful red bird spotted near the lake".to_string()),
                photo_url: None,
                trip_id: None,
                is_shared: true,
            },
        )
        .await
        .expect("Observation creation with coordinates should succeed");

    println!("✓ Observation created: {}", observation.id);
    println!("  Species: {}", observation.species_name);
    println!("  Location: {}", observation.location);
    println!("  Coordinates: ({}, {})", 
             observation.latitude.unwrap(), 
             observation.longitude.unwrap());

    // Verify coordinates are stored correctly
    assert_eq!(observation.latitude, Some(central_park_lat));
    assert_eq!(observation.longitude, Some(central_park_lng));

    println!("\n=== Step 3: Retrieve observation and verify coordinates persist ===");
    let retrieved = observation_service
        .get_by_id(observation.id)
        .await
        .expect("Observation retrieval should succeed");

    assert_eq!(retrieved.latitude, Some(central_park_lat));
    assert_eq!(retrieved.longitude, Some(central_park_lng));
    println!("✓ Coordinates persisted correctly");

    println!("\n=== Step 4: Create observation without coordinates ===");
    let observation_no_coords = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "Blue Jay".to_string(),
                observation_date,
                location: "Somewhere in the park".to_string(),
                latitude: None,
                longitude: None,
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: false,
            },
        )
        .await
        .expect("Observation creation without coordinates should succeed");

    assert_eq!(observation_no_coords.latitude, None);
    assert_eq!(observation_no_coords.longitude, None);
    println!("✓ Observation without coordinates created successfully");

    println!("\n=== Step 5: Test coordinate validation - invalid latitude ===");
    let invalid_lat_result = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "Invalid Bird".to_string(),
                observation_date,
                location: "Invalid location".to_string(),
                latitude: Some(95.0), // Invalid: > 90
                longitude: Some(-73.0),
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: false,
            },
        )
        .await;

    assert!(invalid_lat_result.is_err(), "Invalid latitude should be rejected");
    println!("✓ Invalid latitude rejected: {}", invalid_lat_result.unwrap_err());

    println!("\n=== Step 6: Test coordinate validation - invalid longitude ===");
    let invalid_lng_result = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "Invalid Bird".to_string(),
                observation_date,
                location: "Invalid location".to_string(),
                latitude: Some(40.0),
                longitude: Some(-185.0), // Invalid: < -180
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: false,
            },
        )
        .await;

    assert!(invalid_lng_result.is_err(), "Invalid longitude should be rejected");
    println!("✓ Invalid longitude rejected: {}", invalid_lng_result.unwrap_err());

    println!("\n=== Step 7: Test coordinate validation - incomplete coordinates ===");
    let incomplete_coords_result = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "Invalid Bird".to_string(),
                observation_date,
                location: "Invalid location".to_string(),
                latitude: Some(40.0),
                longitude: None, // Missing longitude
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: false,
            },
        )
        .await;

    assert!(incomplete_coords_result.is_err(), "Incomplete coordinates should be rejected");
    println!("✓ Incomplete coordinates rejected: {}", incomplete_coords_result.unwrap_err());

    println!("\n=== Step 8: Update observation coordinates ===");
    // Times Square coordinates
    let times_square_lat = 40.758896;
    let times_square_lng = -73.985130;

    let updated = observation_service
        .update(
            observation.id,
            user.id,
            UpdateObservationRequest {
                species_name: None,
                observation_date: None,
                location: Some("Times Square, New York".to_string()),
                latitude: Some(times_square_lat),
                longitude: Some(times_square_lng),
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: None,
            },
        )
        .await
        .expect("Observation update should succeed");

    assert_eq!(updated.latitude, Some(times_square_lat));
    assert_eq!(updated.longitude, Some(times_square_lng));
    println!("✓ Coordinates updated successfully");
    println!("  New location: {}", updated.location);
    println!("  New coordinates: ({}, {})", 
             updated.latitude.unwrap(), 
             updated.longitude.unwrap());

    println!("\n=== Step 9: Create additional observations for proximity search ===");
    // Brooklyn Bridge (about 8 km from Times Square)
    let brooklyn_lat = 40.706086;
    let brooklyn_lng = -73.996864;
    
    let obs_brooklyn = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "American Robin".to_string(),
                observation_date,
                location: "Brooklyn Bridge".to_string(),
                latitude: Some(brooklyn_lat),
                longitude: Some(brooklyn_lng),
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: true,
            },
        )
        .await
        .expect("Brooklyn observation should be created");

    println!("✓ Brooklyn Bridge observation created");

    // Statue of Liberty (about 10 km from Times Square)
    let liberty_lat = 40.689247;
    let liberty_lng = -74.044502;
    
    let obs_liberty = observation_service
        .create(
            user.id,
            CreateObservationRequest {
                species_name: "Seagull".to_string(),
                observation_date,
                location: "Statue of Liberty".to_string(),
                latitude: Some(liberty_lat),
                longitude: Some(liberty_lng),
                notes: None,
                photo_url: None,
                trip_id: None,
                is_shared: true,
            },
        )
        .await
        .expect("Liberty observation should be created");

    println!("✓ Statue of Liberty observation created");

    println!("\n=== Step 10: Test proximity search (5 km radius from Times Square) ===");
    let nearby_5km = observation_service
        .find_nearby(
            times_square_lat,
            times_square_lng,
            5.0,
            Some(user.id),
            None,
        )
        .await
        .expect("Proximity search should succeed");

    println!("✓ Found {} observations within 5 km", nearby_5km.len());
    for obs_with_dist in &nearby_5km {
        println!("  - {} at {} km: {}", 
                 obs_with_dist.observation.species_name,
                 obs_with_dist.distance_km,
                 obs_with_dist.observation.location);
        
        // Verify distance is within radius
        assert!(obs_with_dist.distance_km <= 5.0, 
                "Distance {} km exceeds radius 5 km", 
                obs_with_dist.distance_km);
        
        // Verify distance field is present and valid
        assert!(obs_with_dist.distance_km >= 0.0);
        assert!(obs_with_dist.distance_km.is_finite());
    }

    // The updated observation (at Times Square) should be found
    let times_square_found = nearby_5km.iter()
        .any(|o| o.observation.id == observation.id);
    assert!(times_square_found, "Times Square observation should be within 5 km");

    println!("\n=== Step 11: Test proximity search (15 km radius from Times Square) ===");
    let nearby_15km = observation_service
        .find_nearby(
            times_square_lat,
            times_square_lng,
            15.0,
            Some(user.id),
            None,
        )
        .await
        .expect("Proximity search should succeed");

    println!("✓ Found {} observations within 15 km", nearby_15km.len());
    
    // All three observations should be found
    assert!(nearby_15km.len() >= 3, "Should find at least 3 observations within 15 km");
    
    let brooklyn_found = nearby_15km.iter()
        .any(|o| o.observation.id == obs_brooklyn.id);
    assert!(brooklyn_found, "Brooklyn observation should be within 15 km");
    
    let liberty_found = nearby_15km.iter()
        .any(|o| o.observation.id == obs_liberty.id);
    assert!(liberty_found, "Liberty observation should be within 15 km");

    println!("\n=== Step 12: Verify Haversine distance calculations ===");
    let calculated_brooklyn_dist = GeoService::haversine_distance(
        times_square_lat, times_square_lng,
        brooklyn_lat, brooklyn_lng
    );
    println!("✓ Distance to Brooklyn Bridge: {:.2} km", calculated_brooklyn_dist);
    assert!(calculated_brooklyn_dist > 0.0);
    assert!(calculated_brooklyn_dist < 15.0);

    let calculated_liberty_dist = GeoService::haversine_distance(
        times_square_lat, times_square_lng,
        liberty_lat, liberty_lng
    );
    println!("✓ Distance to Statue of Liberty: {:.2} km", calculated_liberty_dist);
    assert!(calculated_liberty_dist > 0.0);
    assert!(calculated_liberty_dist < 15.0);

    // Verify distances in search results match calculated distances
    for obs_with_dist in &nearby_15km {
        let expected_dist = GeoService::haversine_distance(
            times_square_lat, times_square_lng,
            obs_with_dist.observation.latitude.unwrap(),
            obs_with_dist.observation.longitude.unwrap()
        );
        
        let diff = (obs_with_dist.distance_km - expected_dist).abs();
        assert!(diff < 0.01, 
                "Distance mismatch for {}: reported={}, calculated={}, diff={}",
                obs_with_dist.observation.species_name,
                obs_with_dist.distance_km,
                expected_dist,
                diff);
    }
    println!("✓ All distance calculations are accurate");

    println!("\n=== Step 13: Test proximity search with species filter ===");
    let cardinal_nearby = observation_service
        .find_nearby(
            times_square_lat,
            times_square_lng,
            15.0,
            Some(user.id),
            Some("Cardinal".to_string()),
        )
        .await
        .expect("Proximity search with filter should succeed");

    println!("✓ Found {} Cardinal observations within 15 km", cardinal_nearby.len());
    
    // Should find the updated observation (Northern Cardinal)
    for obs_with_dist in &cardinal_nearby {
        assert!(obs_with_dist.observation.species_name.contains("Cardinal"),
                "Filtered results should only contain Cardinals");
    }

    println!("\n=== Step 14: Clean up test data ===");
    cleanup_user(&pool, &username).await;
    println!("✓ Test data cleaned up");

    println!("\n=== ✅ All integration tests passed! ===");
    println!("\nTested features:");
    println!("  ✓ User registration");
    println!("  ✓ Observation creation with coordinates");
    println!("  ✓ Observation creation without coordinates");
    println!("  ✓ Coordinate storage and retrieval");
    println!("  ✓ Coordinate validation (latitude bounds)");
    println!("  ✓ Coordinate validation (longitude bounds)");
    println!("  ✓ Coordinate validation (pair requirement)");
    println!("  ✓ Coordinate updates");
    println!("  ✓ Proximity search with various radii");
    println!("  ✓ Distance calculations (Haversine formula)");
    println!("  ✓ Distance inclusion in search results");
    println!("  ✓ Proximity search with species filter");
}
